text/0000-render-element-modifiers.md

```diff
@@ -0,0 +1,570 @@
+```
+
+Currently, neither modifiers nor helpers in Glimmer are given the context of the
+template at any point. Both the `{{action}}` helper and modifier are given the
+context as an implicit first argument, via an AST transform. The above becomes
+the following in the final template, before it is compiled into the Glimmer byte
+code:
+
+```hbs
+<div {{did-render (action this this.teardownElement)}}></div>
+```
+
+This gives `{{action}}` the correct context to bind the function it is passed
+and was done purely for backwards compatibility, since `{{action}}` existed
+before modifiers and helpers were fully rationalized as features.
+
+Adding this implicit context to other helpers and modifiers would require
+changes to the Glimmer VM and is a much larger language design problem. As such,
+we believe it is out of scope for this RFC. Default binding behavior could be
+added in the future, if a context API is decided on.
+
+## How we teach this
+
+Element modifiers will be new to everyone, so we're starting with a mostly blank
+slate. The only modifier that exists in classic Ember is `{{action}}`, and while
+most existing users will be familiar with it, that familiarity may not translate
+to the more general idea of modifiers.
+
+The first thing we should focus on is teaching _modifiers in general_. Modifiers
+should be seen as the place for any logic which needs to act directly on an
+element, or when an element is added to or removed from the DOM. Modifiers can
+be fully independent (for instance, a `scroll-to` modifier that transparently
+manages the scroll position of the element) or they can interact with the
+component (like the `did-render` and `will-destroy` modifiers). In all cases
+though, they are _tied to the render lifecycle of the element_, and they
+generally contain _side-effects_ (though these may be transparent and
+declarative, as in the case of `{{action}}` or the theoretical `{{scroll-to}}`).
+
+Second, we should teach the render modifiers specifically. We can do this by
+illustrating common use cases which can currently be solved with render hooks,
+and comparing them to using modifiers for the same solution.
+
+One thing we should definitely avoid teaching except in advanced cases is the
+_ordering_ of element modifiers. Ideally, element modifiers should be
+commutative, and order should not be something users have to think about. When
+custom element modifiers become widely available, this should be considered best
+practice.
+
+### Example: Scrolling an element to a position
+
+This sets the scroll position of an element, and updates it whenever the scroll
+position changes.
+
+Before:
+
+```hbs
+{{yield}}
+```
+```js
+export default Component.extend({
+  classNames: ['scroll-container'],
+
+  didRender() {
+    this.element.scrollTop = this.scrollPosition;
+  }
+});
+```
+
+After:
+
+```hbs
+<div {{did-render this.setScrollPosition @scrollPosition}} class="scroll-container">
```

Comment:
This example definitely is correct, but the lack of a `this` binding means that the component method will actually end up being called _unbound_ which would probably be pretty confusing (if that was how we taught it)...